### PR TITLE
[WIP] Full API, multi-stock, multi-channel - Needs splitting.

### PR DESCRIPTION
**Pull request by pjedrzejewski**
Hey folks!

A lot of people know that we have started working on the following things for Sylius:

* Full API
* Multi-Stock
* Multi-Channel

On one of our projects, we needed this, although schedule was a bit crazy. I'd say close to impossible, but we did it! While this implementation is definitely not Sylius Core ready, I think it is a good base for the final implementations of these features.

I hope to get back to this very soon, but there are several people, who are willing to help finishing this, so I decided to share what we have already.

Let me summarize what is done here:

1. API - Pretty much complete REST API for Sylius. While some responses/serialization configs could use some optimization, this implementation allows all things described here - http://docs.sylius.org/en/latest/api/index.html. Yes I got smarter and worked on docs together with the code. :dancer: I think this is the easiest one to finish. If we manage to extract the API related changes from the rest - it is insta-merge.

2. Multi-Channel - For our project, Sylius has been used as backend and via API, so frontend aspect of multi-channel is obviously not ready. While I have made some adjustments to currency selection and so on. There is some work needed to make it functional from the frontend. Display right products and so on.

3. Multi-Stock - I think for this, what needs to be done is hooking this up with Zones system, so that StockLocationProvider takes stock locations which can ship to the particular shipping address. Generally, it needs to be tested and probably a lot will change - it is the most complex part.

**Obviously, I'm aware that this PR is unreviewable it its current form, but there are people in desperate need to these features, so I thought it is better to share it like this, than keep it hidden for next weeks.**

After I deal with few other things, I would like to split this PR into 3 separate branches. Actually I started working on this on 2 separate branches, but as the deadlines were coming ... It got messed up.

I hope to come up with a list of things that need to be done ASAP. For multi-stock, the implementation is heavily inspired by https://guides.spreecommerce.com/developer/inventory.html.

For multi-channel, a ridcolously small amount of docs is available here - http://docs.sylius.org/en/latest/book/channels.html, but I think the idea is quite clear.

### CODE REVIEW

**Comment by kayue** [comment]
Yes we needed this as well!

**Comment by AlucardleVash** [comment]
Translation doesn't work for Taxon, and i don't find why. Maybe someone have an idea ?

File: src/Sylius/Bundle/InventoryBundle/EventListener/StockItemListener.php [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+     */
+    public function __construct(StockItemFactoryInterface $stockItemFactory)
+    {
+        $this->stockItemFactory = $stockItemFactory;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function onStockableCreate(GenericEvent $event)
```
Comment (umpirsky) [inline src/Sylius/Bundle/InventoryBundle/EventListener/StockItemListener.php]: We usually name listener methods by what they do rather then by event names (e.g.: `OrderNumberListener::generateOrderNumber()`, `OrderUpdateListener::recalculateOrderTotal()`...).

File: src/Sylius/Bundle/InventoryBundle/Form/Type/StockLocationType.php [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+            ))
+            ->add('name', 'text', array(
+                'label' => 'sylius.form.stock_location.name'
+            ))
+            ->add('enabled', 'checkbox', array(
+                'required' => false,
+                'label'    => 'sylius.form.stock_location.enabled',
+            ))
+        ;
+
```
Comment (umpirsky) [inline src/Sylius/Bundle/InventoryBundle/Form/Type/StockLocationType.php]: Extra line.

**Comment by kayue** [comment]
Multi-stock is not on master yet...

**Comment by kayue** [comment]
@pjedrzejewski may I know do you have any plan for multi-stock? can we help?

**Comment by lbaey** [comment]
Hey, what's the status on this feature ? Thx

**Comment by michalmarcinkowski** [comment]
Multi channel and API is already on mater. The multi-stock is in progress by @kayue and @umpirsky :)

**Comment by halk** [comment]
Are @kayue and @umpirsky still working on multi-stock?

**Comment by pjedrzejewski** [comment]
@halk Multi-stock will not be part of the core in the 1.0, although adding it is possible - we have a single point of entry for all inventory operations. For example, the service that checks the availability can consider multiple warehouses in your custom implementation. 

We also have released http://docs.sylius.org/en/latest/cookbook/creating-extension.html and will be expanding this guide, so hopefully someone will be interested in implementing it as an extension.

**Comment by adamdyson** [comment]
Can someone provide an update on what is the state of multi-inventory for Sylius?

**Comment by kayue** [comment]
@adamdyson In our own project, we have created our multi-channel (warehouses) stock management system on top of Sylius. It is based on a "Stock Ledger" model that ERPnext.com is using, and very different from what Sylius's previous approach.

**Comment by adamdyson** [comment]
@kayue Care to share?

The company I work for has asked me to evaluate various PHP based eCommerce platforms, and put together a migration plan for our current websites built upon Magento 1.

Multi-stock inventory, warehousing is a must have for us. This can easily be achieved with Magento 2 using either their built in Order Management or by using the many extensions available (free and commercial).

If this is not something Sylius can offer, then I have no choice but to remove them from the list of possible platforms we could potentially use. Which is disappointing.

**Comment by kayue** [comment]
@adamdyson If you must use PHP, take a look at marello.com. I don't think there are many options.

And to be frank, it is not a standard feature for e-commerce. Usually it is the ERP take care of these stuff.

**Comment by umpirsky** [comment]
It would be cool if there is integration with Akeneo PIM, I guess that would solve multi-channel (warehouses) stock management, don't you think?

Quick search listed:

- https://github.com/Sylake/AkeneoProducerBundle
- https://github.com/Sylake

**Comment by adamdyson** [comment]
@kayue I disagree, inventory management is one of the most important features. Any medium-large sized business is going to have multiple warehouses, shops, POS etc. So unless Sylius is only targeting small businesses, it's a must have feature.

https://magento.com/products/order-management
https://www.shopify.com/enterprise/multi-channel-inventory-management
http://build.prestashop.com/news/stock-management-in-prestashop-1-7/
https://codecanyon.net/item/woocommerce-warehouses/13087646

I known an ERP is generally considered a single source of truth, but even with an ERP (which we use) the eCommerce platform still needs to support multi-inventory. Syncing between the two is then up to the developer.

**Comment by jupiter01** [comment]
Any further updates on this?

**Pull request closed without merging**